File: knapsack/c/knapsack.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#include "dynprog.h"

int main(const int argc, char const * const * const argv) {
    if (argc != 2) {
        printf("Usage: knapsack filename\n");
        return 0;
    }
    
    const char* const filename = argv[1];
    FILE* pFile = fopen(filename, "r");
    if (!pFile) {
        fprintf(stderr, "Can't open input file\n");
        exit(1);
    }

    int n;
    int K;
    fscanf(pFile, "%d %d", &n, &K);

    int* values = (int*)malloc(sizeof(int) * n); 
    if (!values) {
        fclose(pFile);

        fprintf(stderr, "Error allocating memory\n");
        exit(1);
    }

    int* weights = (int*)malloc(sizeof(int) * n); 
    if (!weights) {
        fclose(pFile);

        fprintf(stderr, "Error allocating memory\n");
        exit(1);
    }

    for (int i = 0; i < n; i++) {
        fscanf(pFile, "%d %d", &values[i], &weights[i]);
    }

    fclose(pFile);

    int* result = NULL;

    clock_t start = clock();

    int result_size = dynamic_programming(n, K, values, weights, &result);

    clock_t end = clock();
    float seconds = (float)(end - start) / CLOCKS_PER_SEC;

    printf("Total items to take:%d\nItems:\n", result_size);
    for (int i = result_size - 1; i >= 0; --i)
        printf("%d\n", result[i]);
    free(result);

    printf("Completed in %f seconds.\n", seconds);

    free(values);
    free(weights);

    return 0;
}

File: knapsack/python/knapsack.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
import sys
import numpy as np
import time
from dynprog import dynamic_programming


def main(argv):
    if len(argv) != 2:
        print("Usage: knapsack filename")
        return

    filename = argv[1]
    with open(filename) as f:
        line = f.readline()
        n, K = line.split(' ')
        n = int(n)
        K = int(K)

        values = np.zeros((n,), dtype=np.int)
        weights = np.zeros((n,), dtype=np.int)

        for i in range(n):
            line = f.readline()
            v, w = line.split(' ')
            values[i] = int(v)
            weights[i] = int(w)

    start = time.time()
    result = dynamic_programming(n, K, values, weights)
    end = time.time()
    print("Total items to take:{}\nItems:", len(result))
    for i in result:
        print(i)
    print("Completed in {} seconds.".format(end - start))

if __name__ == "__main__":
    main(sys.argv)
